### DISCUSSION TITLE
Missing example for database migrations

### DISCUSSION THREAD

**Discussion by shomodj** [pocketbase/pocketbase#175] [Q&A]
Hi @ganigeorgiev, first I want to thank you for this amazing project, I really like the idea.

Now I have used django migrations before successfully, and I'm a bit struggling here as there is no clear example.

Is the migration code (create/alter/...) created automatically, or do we need to do it manually?

Also, what about the tables that PocketBase creates for itself? How do we add them to the migrations, and what about the data it then enters into those tables.

Anyway my use case would be to start a clean PocketBase instance, create a migration (expecting to see default PocketBase tables), add my tables and then create the migration again and so on. I'm not sure how to do this?

Thank you for this amazing project

**Comment by ganigeorgiev** [comment]
Migrations in PocketBase are available only when you use it as a framework - https://pocketbase.io/docs/db-migrations/.
If future version, I'll consider adding something like `./pocketbase migrate export-collections` command that will generate a migration file for you with the local collections state (currently you have to write the migration file on your own).

Exprort/Import of user defined collections and data is already in the roadmap after the Dart SDK (see #48).
This practically means, that you'll have an option in the Admin UI to export your collections (and/or their data) to a JSON file which later you'll be able to import in another PocketBase instance (eg. running on a remote server) so that the same collections (and/or data) are available there too.

I'll close the issue for now, since the topic is already covered in other places, but if the above is not what you were looking for, please comment and I'll reopen it.

**Comment by shomodj** [comment]
Well, I'm using it as a framework

```
package migrations

import (
    "github.com/pocketbase/dbx"
    m "github.com/pocketbase/pocketbase/migrations"
    "log"
)

func init() {
    m.Register(func(db dbx.Builder) error {
        posts := make(map[string]string)

        posts["id"] = "TEXT PRIMARY KEY"
        posts["created"] = "TEXT DEFAULT '' NOT NULL"
        posts["updated"] = "TEXT DEFAULT '' NOT NULL"
        posts["content"] = "TEXT DEFAULT ''"
        posts["converted"] = "TEXT DEFAULT ''"

        query := db.CreateTable("posts", posts)
        _, err := query.Execute()

        if err != nil {
            log.Fatalln(err)
        }
        
        return nil
    }, func(db dbx.Builder) error {
        // add down queries...

        return nil
    })
}
```
And when I run `migrate up` it does create the table in sqlite but PocketBase does not show it.

So my question is still valid, migration doc is missing a concrete example on how to create tables with code? Or I'm missing something?

**Comment by ganigeorgiev** [comment]
In that case, you'll have to create a collection and save it and it will take care for creating the related records table for you. Or in other words, your migration could look something like this:

```go
package migrations

import (
	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase/daos"
	"github.com/pocketbase/pocketbase/models"
	"github.com/pocketbase/pocketbase/models/schema"
	m "github.com/pocketbase/pocketbase/migrations"
)

func init() {
	m.Register(func(db dbx.Builder) error {
		collection := &models.Collection{
			Name:       "posts",
			CreateRule: nil,
			ListRule:   nil,
			ViewRule:   nil,
			UpdateRule: nil,
			DeleteRule: nil,
			Schema: schema.NewSchema(
				&schema.SchemaField{
					Name:     "content",
					Type:     schema.FieldTypeText,
					Required: true,
					Options:  &schema.TextOptions{},
				},
				&schema.SchemaField{
					Name:    "converted",
					Type:    schema.FieldTypeText,
					Options: &schema.TextOptions{},
				},
				// ... etc.
			),
		}

		return daos.New(db).SaveCollection(collection)
	}, func(db dbx.Builder) error {
		dao := daos.New(db)

		collection, err := dao.FindCollectionByNameOrId("posts")
		if err != nil {
			return err
		}

		return daos.New(db).DeleteCollection(collection)
	})
}
```

This indeed is not documented at the moment. More advanced usage examples and documentation are also planned in the [roadmap](https://github.com/orgs/pocketbase/projects/2), but as I said, we'll have a command that will generate the above for us.

Creating collections programmatically is also mentioned in this discussion - https://github.com/pocketbase/pocketbase/discussions/125

**Comment by ganigeorgiev** [comment]
@shomodj Just to let you know that a new migration command was added in [v0.4.0 release](https://github.com/pocketbase/pocketbase/releases/tag/v0.4.0) that will generate a full snapshot of your current collections confuguration:
```sh
go run main.go migrate collections
```

NB! This will work only for new projects - more info about the existing limitations you could find in the [DB migrations guide](https://pocketbase.io/docs/db-migrations/#collections-migration).

**Reply by shomodj** [reply]
@ganigeorgiev Thank you!
